lib/engine.js

```diff
@@ -835,6 +841,88 @@ module.exports = function (s, conf) {
     }
   }
 
+  function withOnPeriod (trade, period_id) {
+    updatePeriod(trade)
+    if (!s.in_preroll) {
+      if (so.mode !== 'live' && !s.start_capital) {
+        s.start_capital = 0
+        s.start_price = trade.price
+        if (so.asset_capital) {
+          s.start_capital += so.asset_capital * s.start_price
+        }
+        if (so.currency_capital) {
+          s.start_capital += so.currency_capital
+        }
+      }
+      if (!so.manual) {
+        executeStop()
+        if (s.signal) {
+          executeSignal(s.signal)
+          s.signal = null
+        }
+      }
+      if (so.mode !== 'live') {
+        adjustBid(trade)
+        executeOrder(trade)
+      }
+    }
```

Comment:
(out of scope for this PR, I think, but:) I am wondering if we should have a function here for stop loss/profit stop, since I'm guessing we're only doing those on periods now, where it might make more sense to do it on trades. Thoughts?

Reply:
I'd tend to agree. We'll have to approach it a little carefully as a single trade outside the thresholds could trigger the system to initialize a stop order where currently the period has to close outside the threshold. I'd like the stop analysis to at least happen onTrade and then could emit a buy/sell signal on the eventBus independent of other pieces of the system. Having buy/sell signals come over the bus also means that strategies could also emit them from their `calculate` functions if they wanted to.